packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+	         verts_to_recolor_atomic, verts_to_recolor_size_atomic, verts_to_send_atomic, verts_to_send_size_atomic, 
+		 recoloringSize, rand_dev, gid_dev, ghost_degrees_dev, recolor_degrees);
+
+        //copy the updated device views back into host memory where necessary
+        Kokkos::deep_copy(verts_to_send_host, verts_to_send_view);
+        Kokkos::deep_copy(verts_to_send_size_host, verts_to_send_size);
+	//we only use the list of verts to recolor on device, no need to copy to host.
+        Kokkos::deep_copy(verts_to_recolor_size_host, verts_to_recolor_size);
+	Kokkos::deep_copy(recoloringSize_host, recoloringSize);
+        
+	if(distributedRounds < numStatisticRecordingRounds){
+	  conflictDetectionPerRound[distributedRounds] = timer() - detection_temp;
+          conflict_detection += conflictDetectionPerRound[distributedRounds];
+          compPerRound[distributedRounds] += conflictDetectionPerRound[distributedRounds];
+          totalPerRound[distributedRounds] += conflictDetectionPerRound[distributedRounds];
+          comp_time += conflictDetectionPerRound[distributedRounds];
+        }
+
+        distributedRounds++;
+        int localDone = recoloringSize_host(0);
+        int globalDone = 0;
+        Teuchos::reduceAll<int,int>(*comm, Teuchos::REDUCE_SUM, 1, &localDone, &globalDone);
+        done = !globalDone;
+        
+      }//end device coloring
+      
+      if(recoloringSize_host(0) > 0 || !done){
+
+	ghost_colors_host = Kokkos::create_mirror_view(ghost_colors);
+	deep_copy(ghost_colors_host, ghost_colors);
+
+	boundary_verts_host = Kokkos::create_mirror_view(boundary_verts_dev);
+	deep_copy(boundary_verts_host, boundary_verts_dev);
+      }
+
+      while(recoloringSize_host(0) > 0 || !done){
+	auto femvColors_host = femv->getLocalViewHost();
+	auto colors_host = subview(femvColors_host, Kokkos::ALL, 0);
+	if(distributedRounds < numStatisticRecordingRounds){
+	  vertsPerRound[distributedRounds] = recoloringSize_host(0);
+	}
+	if(verbose) std::cout<<comm->getRank()<<": starting to recolor, serial\n";
+        comm->barrier();
+	double recolor_temp = timer();
+	if(verts_to_recolor_size_host(0) > 0){
+	  this->colorInterior_serial(femv_colors.size(), dist_adjs_host, dist_offsets_host, femv, 
+			             verts_to_recolor_host, verts_to_recolor_size(0), true);
+	}
+	if(distributedRounds < numStatisticRecordingRounds){
+	  recoloringPerRound[distributedRounds] = timer() - recolor_temp;
+	  recoloring_time += recoloringPerRound[distributedRounds];
+	  total_time += recoloringPerRound[distributedRounds];
+	  comp_time += recoloringPerRound[distributedRounds];
+	  compPerRound[distributedRounds] = recoloringPerRound[distributedRounds];
+	  totalPerRound[distributedRounds] = recoloringPerRound[distributedRounds];
+	}
+	for(size_t i = 0; i < n_ghosts; i++){
+	  colors_host(i+n_local) = ghost_colors_host(i);
+	}
+        comm->barrier();
```

Comment:
Is barrier needed only for timing?

Reply:
There's a verbose-conditioned barrier inside doOwnedToGhosts, so this barrier is unnecessary. I removed it.